quic/s2n-quic-core/src/datagram/default.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// s2n-quic's default implementation of the datagram component
+
+use crate::{
+    datagram::{ConnectionInfo, Endpoint, Packet, Receiver, Sender},
+    event::Timestamp,
+};
+use alloc::collections::VecDeque;
+use bytes::Bytes;
+use core::time::Duration;
+
+#[derive(Debug, Default)]
+pub struct Disabled;
+
+impl Endpoint for Disabled {
+    type Sender = DisabledSender;
+    type Receiver = DisabledReceiver;
+
+    fn create_connection(&mut self, _info: &ConnectionInfo) -> (Self::Sender, Self::Receiver) {
+        (DisabledSender, DisabledReceiver)
+    }
+}
+
+pub struct DisabledSender;
+pub struct DisabledReceiver;
+
+impl Sender for DisabledSender {
+    fn on_transmit<P: Packet>(&mut self, _packet: &mut P) {}
+
+    #[inline]
+    fn has_transmission_interest(&self) -> bool {
+        false
+    }
+}
+
+impl Receiver for DisabledReceiver {
+    fn on_datagram(&self, _datagram: &[u8]) {}
+}
+
+pub struct DefaultSender {
+    queue: VecDeque<Datagram>,
+    datagram_expiration: Duration,
+}
+
+struct Datagram {
+    data: Bytes,
+    creation_time: Option<Timestamp>,
+}
+
+impl Sender for DefaultSender {
+    fn on_transmit<P: Packet>(&mut self, packet: &mut P) {
+        // Cede space to stream data when datagrams are not prioritized
+        if packet.has_pending_streams() && !packet.datagrams_prioritized() {
+            return;
+        }
+
+        while packet.remaining_capacity() > 0 {
+            if let Some(datagram) = self.queue.pop_front() {
+                // Ensure datagram is not too old
+                if let Some(creation_time) = datagram.creation_time {
```

Comment:
We need to have a way to put the creation time on datagrams if it hasn't already been set. Otherwise this code won't ever be executed.